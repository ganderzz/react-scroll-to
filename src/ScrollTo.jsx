import React, { Component, createContext } from "react";

export const ScrollToContext = createContext("scrollToContext");

/**
 * Component that uses render props to inject
 * a function that allows the consumer to scroll to a
 * position in the window or ScrollArea component
 */
class ScrollTo extends Component {
  constructor(props) {
    super(props);

    this.scrollArea = {};

    this.getContext = {
      addScrollArea: (id, ref) => {
        this.scrollArea[id] = ref;
      },
      removeScrollArea: id => {
        delete this.scrollArea[id];
      }
    };
  }

  handleScroll = (props = {}) => {
    const scrollAreaKeys = Object.keys(this.scrollArea);
    const { id, ref, ...rest } = props;

    if (ref) {
      const refNode = ref.current ? ref.current : ref;

      // Scroll by ref
      this._scrollNode(refNode, rest);
    } else if (id) {
      // Scroll by id
      const node = this.scrollArea[id];

      this._scrollNode(node, rest);
    } else if (scrollAreaKeys.length > 0) {
      // Scroll by all scroll areas
      scrollAreaKeys.forEach(key => {
        const node = this.scrollArea[key];

        this._scrollNode(node, rest);
      });
    } else {
      // Scroll by window
      this._scrollNode(window, rest);
    }
  };

  _scrollNode = (node, options) => {
    if (!node) {
      return;
    }

    const top = options.y;
    const left = options.x;

    if (node.scrollTo) {
      node.scrollTo({
        top,
        left,
        behavior: options.smooth ? "smooth" : "auto"
      });
    } else {
      node.scrollLeft = left;
      node.scrollTop = top;
    }
  };

  render() {
    return (
      <ScrollToContext.Provider value={this.getContext}>
        {this.props.children &&
          this.props.children({
            scrollTo: this.handleScroll
          })}
      </ScrollToContext.Provider>
    );
  }
}

ScrollTo.defaultProps = {
  children: () => {}
};

export default ScrollTo;
