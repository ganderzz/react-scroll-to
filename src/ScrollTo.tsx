import * as React from "react";
import { IScrollOptions } from "./IScrollOptions";
import { scrollNode } from "./logic";
import { useScrollTo } from "./UseScrollTo";

interface IProps {
  children?: (props: {
    scroll: (props?: IScrollOptions) => void;
  }) => JSX.Element;
}

/**
 * Component that uses render props to inject
 * a function that allows the consumer to scroll to a
 * position in the window or ScrollArea component
 */
export function ScrollTo({ children }: IProps) {
  const { scroll } = useScrollTo();

  if (!children) {
    return null;
  }

  return children({ scroll });
}
