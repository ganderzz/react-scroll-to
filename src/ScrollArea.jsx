import React, { Component, createRef } from "react";
import { ScrollToContext } from "./ScrollTo";
import generateId from "./utilities/generateId";

export default class ScrollArea extends Component {
  static contextType = ScrollToContext;

  constructor(props) {
    super(props);

    this.node = createRef();
    this.id = this.props.id || generateId();
  }

  componentDidMount() {
    this.context.addScrollArea(this.id, this.node.current);
  }

  componentWillUnmount() {
    this.context.removeScrollArea(this.id);
  }

  render() {
    const { children, ...props } = this.props;

    return (
      <div {...props} ref={this.node}>
        {children}
      </div>
    );
  }
}
