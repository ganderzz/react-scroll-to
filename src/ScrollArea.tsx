import * as React from "react";
import { ScrollToContext } from "./ScrollTo";
import { generateId } from "./utilities";

interface IProps {
  id: string;
  addScrollArea(id: string, node: unknown);
  removeScrollArea(id: string);
}

export class ScrollArea extends React.Component<IProps> {
  node = React.createRef<HTMLDivElement>();
  id = this.props.id || generateId();

  componentDidMount() {
    this.props.addScrollArea(this.id, this.node.current);
  }

  componentWillUnmount() {
    this.props.removeScrollArea(this.id);
  }

  render() {
    const { children, removeScrollArea, addScrollArea, ...props } = this.props;

    return (
      <div {...props} ref={this.node}>
        {children}
      </div>
    );
  }
}

export default props => (
  <ScrollToContext.Consumer>
    {({ addScrollArea, removeScrollArea }) => (
      <ScrollArea
        {...props}
        removeScrollArea={removeScrollArea}
        addScrollArea={addScrollArea}
      />
    )}
  </ScrollToContext.Consumer>
);
