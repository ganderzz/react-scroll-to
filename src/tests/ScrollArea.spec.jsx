import React from "react";
import { render } from "react-testing-library";
import { ScrollToContext } from "../ScrollTo";
import ScrollArea from "../ScrollArea";

jest.mock("../utilities/generateId", () => () => "mock-id");

const BaseScrollArea = props => {
  const { addScrollArea, removeScrollArea, ...rest } = props;

  return (
    <ScrollToContext.Provider value={{ addScrollArea, removeScrollArea }}>
      <ScrollArea {...rest} />
    </ScrollToContext.Provider>
  );
};

describe("Test Scroll Area.", () => {
  it("should call addScrollArea when mounting.", () => {
    const addScrollArea = jest.fn();
    render(
      <BaseScrollArea addScrollArea={addScrollArea} removeScrollArea={() => {}}>
        <h1>Test</h1>
      </BaseScrollArea>
    );

    expect(addScrollArea).toHaveBeenCalledTimes(1);
    expect(addScrollArea.mock.calls[0]).toMatchSnapshot();
  });

  it("should call removeScrollArea when unmounting.", () => {
    const removeScrollArea = jest.fn();
    const { unmount } = render(
      <BaseScrollArea
        addScrollArea={() => {}}
        removeScrollArea={removeScrollArea}
      >
        <h1>Test</h1>
      </BaseScrollArea>
    );

    unmount();

    expect(removeScrollArea).toHaveBeenCalledTimes(1);
    expect(removeScrollArea.mock.calls[0]).toMatchSnapshot();
  });

  it("should render correctly.", () => {
    const { container } = render(
      <BaseScrollArea
        className="foo"
        addScrollArea={() => {}}
        removeScrollArea={() => {}}
      >
        <h1>Test</h1>
      </BaseScrollArea>
    );

    expect(container).toMatchSnapshot();
  });

  // Used to make coverage 100%
  it("should render default context.", () => {
    const fns = {
      addScrollArea: jest.fn(),
      removeScrollArea: jest.fn()
    };

    const { container, debug } = render(
      <ScrollToContext.Provider value={fns}>
        <ScrollArea style={{ padding: 20 }}>test</ScrollArea>
      </ScrollToContext.Provider>
    );
  });
});
