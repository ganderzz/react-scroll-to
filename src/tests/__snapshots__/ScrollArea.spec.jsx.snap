// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Scroll Area. should call addScrollArea when mounting. 1`] = `
Array [
  "mock-id",
  <div>
    <h1>
      Test
    </h1>
  </div>,
]
`;

exports[`Test Scroll Area. should call removeScrollArea when unmounting. 1`] = `
Array [
  "mock-id",
]
`;

exports[`Test Scroll Area. should render correctly. 1`] = `
<div>
  <div
    class="foo"
  >
    <h1>
      Test
    </h1>
  </div>
</div>
`;
